### PR TITLE
Implementation of fetching AE max columns

### PR DESCRIPTION
**Pull request by v-kaywon**

### CODE REVIEW

**Comment by v-kaywon** [comment]
Performance measurement for fetching varchar(max) columns with and without AE:

Performance test information:
Driver: PDO_SQLSRV
Table columns: char(5), char(5, deterministic encrypted), char(5, randomize encrypted), varchar(max), varchar(max, deterministic encrypted), varchar(max, randomize encrypted)
Number of rows in table: 2
Number of iterations: 1000
Number of measurements: 10 (each measurement is the sum of all 1000 iterations)

Mean(with AE): 98.71267
Mean(without AE): 85.88191

**Comment by Hadis-Fard** [comment]
@meet-bhagdev we have the performance measurement on AE retrieval, can you review?

File: source/shared/core_results.cpp [hunk 1]
```diff
@@ -1563,6 +1563,19 @@ SQLPOINTER read_lob_field( _Inout_ sqlsrv_stmt* stmt, _In_ SQLUSMALLINT field_in
...
 
+        if ( stmt->conn->ce_option.enabled == true ) {
+            // cursor type SQLSRV_CURSOR_BUFFERED has to be FORWARD_ONLY
+            // thus has to close and reopen cursor to reset the cursor buffer
+            core::SQLCloseCursor(stmt);
+            core::SQLExecute(stmt);
+            // FETCH_NEXT until the cursor reaches the row that it was at
+            for (int i = 0; i <= row_count; i++) {
+                core::SQLFetchScroll(stmt, SQL_FETCH_NEXT, 0);
+            }
```
Comment (yitam) [inline source/shared/core_results.cpp]: Why not use the `offset` argument for SQLFetchScroll()?
Also, consider renaming input parameter `row_count` to something like `rows_read`

Comment (v-kaywon) [inline source/shared/core_results.cpp]: FORWARD_ONLY cursor ignores the offset argument. SQLSRV_CURSOR_BUFFERED has to be FORWARD_ONLY.

File: source/shared/core_stmt.cpp [hunk 2]
```diff
@@ -796,6 +798,9 @@ bool core_sqlsrv_fetch( _Inout_ sqlsrv_stmt* stmt, _In_ SQLSMALLINT fetch_orient
         // move to the record requested.  For absolute records, we use a 0 based offset, so +1 since 
         // SQLFetchScroll uses a 1 based offset, otherwise for relative, just use the fetch_offset provided.
         SQLRETURN r = stmt->current_results->fetch( fetch_orientation, ( fetch_orientation == SQL_FETCH_RELATIVE ) ? fetch_offset : fetch_offset + 1 TSRMLS_CC );
+        if ( stmt->cursor_type == SQL_CURSOR_FORWARD_ONLY && stmt->conn->ce_option.enabled == true ) {
+            stmt->fwd_row_index++;
+        }
```
Comment (yitam) [inline source/shared/core_stmt.cpp]: Do we reset this `fwd_row_index` at some point? For example, when `past_fetch_end` is true? Or is it not necessary?

File: source/shared/core_stmt.cpp [hunk 3]
```diff
@@ -2171,6 +2176,7 @@ void get_field_as_string( _Inout_ sqlsrv_stmt* stmt, _In_ SQLUSMALLINT field_ind
 		else {
 			// Get the SQL type of the field. unixODBC 2.3.1 requires wide calls to support pooling
 			core::SQLColAttributeW( stmt, field_index + 1, SQL_DESC_CONCISE_TYPE, NULL, 0, NULL, &sql_field_type TSRMLS_CC );
+            SQLLEN sql_field_len = 0;
```
Comment (yitam) [inline source/shared/core_stmt.cpp]: Probably `sql_field_len` is not used after this?

File: source/shared/core_stmt.cpp [hunk 4]
```diff
@@ -2218,145 +2223,195 @@ void get_field_as_string( _Inout_ sqlsrv_stmt* stmt, _In_ SQLUSMALLINT field_ind
...
-
-						} while( r == SQL_SUCCESS_WITH_INFO && is_truncated_warning( state ));
-					}
-					else {
-
-						// We got the field_len_temp from SQLGetData call.
-						field_value_temp = static_cast<char*>( sqlsrv_realloc( field_value_temp, field_len_temp + extra + 1 ));
-
-						// We have already recieved INITIAL_FIELD_STRING_LEN size data.
-						field_len_temp -= INITIAL_FIELD_STRING_LEN;
```
Comment (yitam) [inline source/shared/core_stmt.cpp]: Looks like the block (lines 2261 to 2269) is the same as the code in the else block (lines 2274 to 2282)?
If so, is it possible to refactor the code?

**Comment by meet-bhagdev** [comment]
@Hadis-Fard Sure! Let me know where the report is and i will take a look.

File: source/shared/core_stmt.cpp [hunk 5]
```diff
@@ -2218,145 +2223,195 @@ void get_field_as_string( _Inout_ sqlsrv_stmt* stmt, _In_ SQLUSMALLINT field_ind
                 if( is_truncated_warning( state ) ) {
 #endif // !_WIN32
```
Comment (yitam) [inline source/shared/core_stmt.cpp]: will we get truncated warning if a client buffer is used, (the data has already been fetched beforehand), would it ever reach this part?
similarly, few lines after this, checking if AE is enabled, can we assume client buffer is not used?

File: source/shared/core_stmt.cpp [hunk 6]
```diff
@@ -2218,145 +2223,195 @@ void get_field_as_string( _Inout_ sqlsrv_stmt* stmt, _In_ SQLUSMALLINT field_ind
...
+                            if( r == SQL_SUCCESS_WITH_INFO ) {
+                                core::SQLGetDiagField( stmt, 1, SQL_DIAG_SQLSTATE, state, SQL_SQLSTATE_BUFSIZE, &len
+                                                       TSRMLS_CC );
+                            }
+
+                        } while( r == SQL_SUCCESS_WITH_INFO && is_truncated_warning( state ));
+                    }
+                    else {
+
+                        // We got the field_len_temp from SQLGetData call.
```
Comment (yitam) [inline source/shared/core_stmt.cpp]: Please add a comment here what's this else block is (if not blah blah ?? )

File: source/shared/core_stmt.cpp [hunk 7]
```diff
@@ -2218,145 +2223,195 @@ void get_field_as_string( _Inout_ sqlsrv_stmt* stmt, _In_ SQLUSMALLINT field_ind
...
+                                core::SQLExecute( stmt );
+                                // FETCH_NEXT until the cursor reaches the row that it was at
+                                for ( int i = 0; i <= stmt->fwd_row_index; i++ ) {
+                                    core::SQLFetchScroll( stmt->current_results->odbc, SQL_FETCH_NEXT, 0 );
+                                }
+                            }
+                            else {
+                                core::SQLFetchScroll( stmt->current_results->odbc, SQL_FETCH_NEXT, 0 );
+                                core::SQLFetchScroll( stmt->current_results->odbc, SQL_FETCH_PRIOR, 0 );
+                            }
```
Comment (yitam) [inline source/shared/core_stmt.cpp]: Please investigate whether it's possible to refactor the above specific code for AE's max fields

File: source/shared/core_results.cpp (hunk 1 above)
Comment (yitam) [inline source/shared/core_results.cpp]: Please investigate the feasibility of not using FORWARD_ONLY when AE is on

**Comment by Hadis-Fard** [comment]
@meet-bhagdev It's mixed in the discussion thread, pasting it here:

Performance measurement for fetching varchar(max) columns with and without AE:

Performance test information:
Driver: PDO_SQLSRV
Table columns: char(5), char(5, deterministic encrypted), char(5, randomize encrypted), varchar(max), varchar(max, deterministic encrypted), varchar(max, randomize encrypted)
Number of rows in table: 2
Number of iterations: 1000
Number of measurements: 10 (each measurement is the sum of all 1000 iterations)

Mean(with AE): 98.71267s (SD = 19.78386)
Mean(without AE): 85.88191s (SD = 9.94258)

File: source/shared/core_results.cpp (hunk 1 above)
Comment (v-kaywon) [inline source/shared/core_results.cpp]: I've double checked. I get an error "Protocol error in TDS stream" when trying to fetch an encrypted max type column with non FORWARD_ONLY cursor (CURSOR_STATIC AND CURSOR_DYNAMIC). https://github.com/Microsoft/msphpsql/blob/dev/source/shared/core_stmt.cpp#L1126 is where the driver sets the cursor to SQL_CURSOR_FORWARD_ONLY when SQLSRV_CURSOR_BUFFERED is on.

Comment (yitam) [inline source/shared/core_results.cpp]: Thanks for checking.

File: source/shared/core_stmt.cpp [hunk 8]
```diff
@@ -796,6 +798,9 @@ bool core_sqlsrv_fetch( _Inout_ sqlsrv_stmt* stmt, _In_ SQLSMALLINT fetch_orient
         // move to the record requested.  For absolute records, we use a 0 based offset, so +1 since 
         // SQLFetchScroll uses a 1 based offset, otherwise for relative, just use the fetch_offset provided.
         SQLRETURN r = stmt->current_results->fetch( fetch_orientation, ( fetch_orientation == SQL_FETCH_RELATIVE ) ? fetch_offset : fetch_offset + 1 TSRMLS_CC );
+        if ( stmt->cursor_type == SQL_CURSOR_FORWARD_ONLY && stmt->conn->ce_option.enabled == true ) {
```
Comment (v-kaywon) [inline source/shared/core_stmt.cpp]: even is the column itself is not encrypted, it still goes into this part of the code. Need further investigation to see if we can determine the encryption status of a column in a result set.

File: source/shared/core_stmt.cpp (hunk 2 above)
Comment (v-kaywon) [inline source/shared/core_stmt.cpp]: It's not necessary to reset fwd_row_index explicitly. past_fetch_end is set to true so the driver can raise an error when the user tries to fetch again even after SQL_NO_DATA. past_fetch_end and fwd_row_index is reset when the statement is re-executed and a new result set is created (https://github.com/v-kaywon/msphpsql/blob/22b4817f658a16edb8d24eba7eba2b2e44e2b01c/source/shared/core_stmt.cpp#L217)

File: source/shared/core_stmt.cpp (hunk 5 above)
Comment (v-kaywon) [inline source/shared/core_stmt.cpp]: we do get truncated warning (the driver try to mimic the same behavior as server cursor). The new commit doesn't check for client buffer anymore.

File: source/shared/core_stmt.cpp [hunk 9]
```diff
@@ -796,6 +799,13 @@ bool core_sqlsrv_fetch( _Inout_ sqlsrv_stmt* stmt, _In_ SQLSMALLINT fetch_orient
         // move to the record requested.  For absolute records, we use a 0 based offset, so +1 since 
         // SQLFetchScroll uses a 1 based offset, otherwise for relative, just use the fetch_offset provided.
         SQLRETURN r = stmt->current_results->fetch( fetch_orientation, ( fetch_orientation == SQL_FETCH_RELATIVE ) ? fetch_offset : fetch_offset + 1 TSRMLS_CC );
+        
+        // when AE is on, need to store the current row number being fetched so a cursor can be reset back to it's original position when getting stream data
+        // fwd_row_index is used to store the current row index, which is incremented everytime SQLFetchScroll is called with SQLSRV_CURSOR_FORWARD_ONLY
```
Comment (yitam) [inline source/shared/core_stmt.cpp]: When AE is *enabled*, will keep track of the numbers of rows being fetched *so far such that* ... back to *its* original position...

**Pull request merged**